File: src/LLM.cpp
#include "LLM.h"

std::atomic_flag sigint_terminating = ATOMIC_FLAG_INIT;

void llm_sigint_signal_handler(int sig)
{
    if (sigint_terminating.test_and_set())
    {
        // in case it hangs, we can force terminate the server by hitting Ctrl+C twice
        // this is for better developer experience, we can remove when the server is stable enough
        fprintf(stderr, "Received second interrupt, terminating immediately.\n");
        exit(1);
    }

    for (auto *inst : LLMProviderRegistry::instance().get_instances())
    {
        inst->stop();
        inst->stop_server();
    }
}

// Use a function to ensure the setup only happens once across all libraries
void ensure_error_handlers_initialized()
{
    if (!LLMProviderRegistry::initialised)
    {
        static std::once_flag initialized;
        std::call_once(initialized, []()
                       {
            set_error_handlers();
            register_sigint_hook(llm_sigint_signal_handler); });
    }
}

LLMProviderRegistry *LLMProviderRegistry::custom_instance_ = nullptr;
bool LLMProviderRegistry::initialised = false;

LLMProvider::~LLMProvider() {}

//=========================== Helpers ===========================//

std::string LLM::LLM_args_to_command(const std::string &model_path, int num_slots, int num_threads, int num_GPU_layers, bool flash_attention, int context_size, int batch_size, bool embedding_only, const std::vector<std::string> &lora_paths)
{
    std::string command = "-m " + model_path +
                          " -t " + std::to_string(num_threads) +
                          " -np " + std::to_string(num_slots) +
                          " -c " + std::to_string(context_size) +
                          " -b " + std::to_string(batch_size);

    if (num_GPU_layers > 0)
        command += " -ngl " + std::to_string(num_GPU_layers);
    if (flash_attention)
        command += " --flash-attn";
    if (embedding_only)
        command += " --embedding";
    for (const auto &lora_path : lora_paths)
        command += " --lora " + lora_path;
    return command;
}

bool LLM::has_gpu_layers(const std::string &command)
{
    std::istringstream iss(command);
    std::vector<std::string> args;
    std::string token;

    // Simple splitting (does not handle quoted args)
    while (iss >> token)
    {
        args.push_back(token);
    }

    for (size_t i = 0; i < args.size(); ++i)
    {
        const std::string &arg = args[i];

        // Match separate argument + value
        if (arg == "-ngl" || arg == "--gpu-layers" || arg == "--n-gpu-layers")
        {
            if (i + 1 < args.size())
            {
                try
                {
                    int val = std::stoi(args[i + 1]);
                    return val > 0;
                }
                catch (...)
                {
                    continue;
                }
            }
        }

        // Match inline --flag=value
        size_t eqPos = arg.find('=');
        if (eqPos != std::string::npos)
        {
            std::string key = arg.substr(0, eqPos);
            std::string value = arg.substr(eqPos + 1);

            if (key == "-ngl" || key == "--gpu-layers" || key == "--n-gpu-layers")
            {
                try
                {
                    int val = std::stoi(value);
                    return val > 0;
                }
                catch (...)
                {
                    continue;
                }
            }
        }
    }

    return false;
}

//=========================== Apply Template ===========================//

json LLM::build_apply_template_json(const json &messages)
{
    json j;
    j["messages"] = messages;
    return j;
}

std::string LLM::parse_apply_template_json(const json &result)
{
    try
    {
        return result.at("prompt").get<std::string>();
    }
    catch (const std::exception &)
    {
    }
    return "";
}

//=========================== Tokenize ===========================//

json LLM::build_tokenize_json(const std::string &query)
{
    json j;
    j["content"] = query;
    return j;
}

std::vector<int> LLM::parse_tokenize_json(const json &result)
{
    try
    {
        return result.at("tokens").get<std::vector<int>>();
    }
    catch (const std::exception &)
    {
    }
    return {};
}

//=========================== Detokenize ===========================//

json LLM::build_detokenize_json(const std::vector<int32_t> &tokens)
{
    json j;
    j["tokens"] = tokens;
    return j;
}

std::string LLM::parse_detokenize_json(const json &result)
{
    try
    {
        return result.at("content").get<std::string>();
    }
    catch (const std::exception &)
    {
    }
    return "";
}

//=========================== Embeddings ===========================//

json LLM::build_embeddings_json(const std::string &query)
{
    json j;
    j["content"] = query;
    return j;
}

std::vector<float> LLM::parse_embeddings_json(const json &result)
{
    try
    {
        return result.at("embedding").get<std::vector<float>>();
    }
    catch (const std::exception &)
    {
    }
    return {};
}

//=========================== Completion ===========================//

json LLM::build_completion_json(const std::string &prompt, int id_slot)
{
    json j;
    j["prompt"] = prompt;
    j["id_slot"] = id_slot;
    j["n_keep"] = n_keep;

    if (!grammar.empty())
    {
        try
        {
            j["json_schema"] = json::parse(grammar);
        }
        catch (const json::parse_error &)
        {
            j["grammar"] = grammar;
        }
    }

    if (completion_params.is_object())
    {
        for (json::const_iterator it = completion_params.begin(); it != completion_params.end(); ++it)
        {
            j[it.key()] = it.value();
        }
    }
    return j;
}

std::string LLM::parse_completion_json(const json &result)
{
    try
    {
        return result.at("content").get<std::string>();
    }
    catch (const std::exception &)
    {
    }
    return "";
}

std::string LLM::completion(const std::string &prompt, CharArrayFn callback, int id_slot, bool return_response_json)
{
    std::string response = completion_json(
        build_completion_json(prompt, id_slot),
        callback,
        false);
    if (return_response_json)
        return response;
    return parse_completion_json(json::parse(response));
}

//=========================== Slot Action ===========================//

json LLMLocal::build_slot_json(int id_slot, const std::string &action, const std::string &filepath)
{
    json j;
    j["id_slot"] = id_slot;
    j["action"] = action;
    j["filepath"] = filepath;
    return j;
}

std::string LLMLocal::parse_slot_json(const json &result)
{
    try
    {
        return result.at("filename").get<std::string>();
    }
    catch (const std::exception &)
    {
    }
    return "";
}

//=========================== Logging ===========================//

void LLMProvider::logging_stop()
{
    logging_callback(nullptr);
}

//=========================== Set Template ===========================//

json LLMProvider::build_set_template_json(std::string chat_template)
{
    json j;
    j["chat_template"] = chat_template;
    return j;
}

//=========================== Lora Adapters Apply ===========================//

json LLMProvider::build_lora_weight_json(const std::vector<LoraIdScale> &loras)
{
    json j = json::array();
    for (const auto &lora : loras)
    {
        j.push_back({{"id", lora.id},
                     {"scale", lora.scale}});
    }
    return j;
}

bool LLMProvider::parse_lora_weight_json(const json &result)
{
    try
    {
        return result.at("success").get<bool>();
    }
    catch (const std::exception &)
    {
    }
    return false;
}

//=========================== Lora Adapters List ===========================//

std::vector<LoraIdScalePath> LLMProvider::parse_lora_list_json(const json &result)
{
    std::vector<LoraIdScalePath> loras;
    try
    {
        for (const auto &lora : result)
        {
            loras.push_back({lora["id"].get<int>(),
                             lora["scale"].get<float>(),
                             lora["path"].get<std::string>()});
        }
    }
    catch (const std::exception &)
    {
    }
    return loras;
}

//=========================== API ===========================//

bool Has_GPU_Layers(const char *command)
{
    return LLM::has_gpu_layers(command);
}

void LLM_Debug(int debug_level)
{
    LLMProviderRegistry &registry = LLMProviderRegistry::instance();
    registry.set_debug_level(debug_level);
    for (auto *inst : registry.get_instances())
    {
        inst->debug(debug_level);
    }
}

void LLM_Logging_Callback(CharArrayFn callback)
{
    LLMProviderRegistry &registry = LLMProviderRegistry::instance();
    registry.set_log_callback(callback);
    for (auto *inst : registry.get_instances())
    {
        inst->logging_callback(callback);
    }
}

void LLM_Logging_Stop()
{
    LLM_Logging_Callback(nullptr);
}

#ifdef _DEBUG
const bool IsDebuggerAttached(void)
{
#ifdef _MSC_VER
    return ::IsDebuggerPresent();
#elif __APPLE__
    return AmIBeingDebugged();
#elif __linux__
    return debuggerIsAttached();
#else
    return false;
#endif
}
#endif

const char *LLM_Tokenize(LLM *llm, const char *query)
{
    json result = llm->tokenize(query);
    return stringToCharArray(result.dump());
}

const char *LLM_Detokenize(LLM *llm, const char *tokens_as_json)
{
    return stringToCharArray(llm->detokenize(json::parse(tokens_as_json)));
}

const char *LLM_Embeddings(LLM *llm, const char *query)
{
    json result = llm->embeddings(query);
    return stringToCharArray(result.dump());
}

const char *LLM_Completion(LLM *llm, const char *prompt, CharArrayFn callback, int id_slot, bool return_response_json)
{
    return stringToCharArray(llm->completion(prompt, callback, id_slot, return_response_json));
}

void LLM_Set_Completion_Parameters(LLM *llm, const char *params_json)
{
    json params = json::parse(params_json ? params_json : "{}");
    llm->set_completion_params(params);
}

const char *LLM_Get_Completion_Parameters(LLM *llm)
{
    return stringToCharArray((llm->completion_params).dump());
}

void LLM_Set_Grammar(LLM *llm, const char *grammar)
{
    llm->set_grammar(grammar);
}

const char *LLM_Get_Grammar(LLM *llm)
{
    return stringToCharArray(llm->grammar);
}

const char *LLM_Get_Template(LLM *llm)
{
    return stringToCharArray(llm->get_template());
}

const char *LLM_Apply_Template(LLM *llm, const char *messages_as_json)
{
    return stringToCharArray(llm->apply_template(json::parse(messages_as_json)));
}

void LLM_Set_Template(LLMProvider *llm, const char *chat_template)
{
    llm->set_template(chat_template);
}

void LLM_Enable_Reasoning(LLMProvider *llm, bool enable_reasoning)
{
    llm->enable_reasoning(enable_reasoning);
}

const char *LLM_Save_Slot(LLMLocal *llm, int id_slot, const char *filepath)
{
    return stringToCharArray(llm->save_slot(id_slot, filepath));
}

const char *LLM_Load_Slot(LLMLocal *llm, int id_slot, const char *filepath)
{
    return stringToCharArray(llm->load_slot(id_slot, filepath));
}

void LLM_Cancel(LLMLocal *llm, int id_slot)
{
    llm->cancel(id_slot);
}

bool LLM_Lora_Weight(LLMProvider *llm, const char *loras_as_json)
{
    try
    {
        json loras_arr = json::array();
        loras_arr = json::parse(loras_as_json);
        std::vector<LoraIdScale> loras;
        for (const auto &lora : loras_arr)
        {
            loras.push_back({lora["id"].get<int>(), lora["scale"].get<float>()});
        }
        return llm->lora_weight(loras);
    }
    catch (const std::exception &)
    {
    }
    return false;
}

const char *LLM_Lora_List(LLMProvider *llm)
{
    std::vector<LoraIdScalePath> loras = llm->lora_list();
    json j = json::array();
    for (const auto &lora : loras)
    {
        j.push_back({{"id", lora.id},
                     {"scale", lora.scale}});
    }
    return stringToCharArray(j.dump());
}

void LLM_Delete(LLMProvider *llm)
{
    if (llm != nullptr)
    {
        delete llm;
    }
}

void LLM_Start_Server(LLMProvider *llm, const char *host, int port, const char *API_key)
{
    llm->start_server(host, port, API_key);
}

void LLM_Stop_Server(LLMProvider *llm)
{
    llm->stop_server();
}

void LLM_Join_Service(LLMProvider *llm)
{
    llm->join_service();
}

void LLM_Join_Server(LLMProvider *llm)
{
    llm->join_server();
}

void LLM_Start(LLMProvider *llm)
{
    llm->start();
}

const bool LLM_Started(LLMProvider *llm)
{
    return llm->started();
}

void LLM_Stop(LLMProvider *llm)
{
    llm->stop();
}

void LLM_Set_SSL(LLMProvider *llm, const char *SSL_cert, const char *SSL_key)
{
    llm->set_SSL(SSL_cert, SSL_key);
}

const int LLM_Status_Code()
{
    return get_status_code();
}

const char *LLM_Status_Message()
{
    std::string result = get_status_message();
    return stringToCharArray(result);
}

const int LLM_Embedding_Size(LLMProvider *llm)
{
    return llm->embedding_size();
}

File: csharp/LLM.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Runtime.InteropServices;
using System.Threading.Tasks;
using Newtonsoft.Json.Linq;

namespace UndreamAI.LlamaLib
{

    // Data structures for LoRA operations
    public struct LoraIdScale
    {
        public int Id { get; set; }
        public float Scale { get; set; }
        
        public LoraIdScale(int id, float scale)
        {
            Id = id;
            Scale = scale;
        }
    }

    public struct LoraIdScalePath
    {
        public int Id { get; set; }
        public float Scale { get; set; }
        public string Path { get; set; }
        
        public LoraIdScalePath(int id, float scale, string path)
        {
            Id = id;
            Scale = scale;
            Path = path;
        }
    }

    // Base LLM class
    public abstract class LLM : IDisposable
    {
        public LlamaLib llamaLib = null;
        public IntPtr llm = IntPtr.Zero;
        protected readonly object _disposeLock = new object();
        public bool disposed = false;
        public int seed = 0;
        public int numPredict = -1;
        public int numKeep = 0;
        public float temperature = 0.80f;
        public string jsonSchema = "";
        public string grammar = "";

        protected LLM() { }

        protected LLM(LlamaLib llamaLibInstance)
        {
            llamaLib = llamaLibInstance ?? throw new ArgumentNullException(nameof(llamaLibInstance));
        }

        public static void Debug(int debugLevel)
        {
            LlamaLib.Debug(debugLevel);
        }

        public static void LoggingCallback(LlamaLib.CharArrayCallback callback)
        {
            LlamaLib.LoggingCallback(callback);
        }

        public static void LoggingStop()
        {
            LlamaLib.LoggingStop();
        }

        protected void CheckLlamaLib()
        {
            if (disposed) throw new ObjectDisposedException(GetType().Name);
            if (llamaLib == null) throw new InvalidOperationException("LlamaLib instance is not initialized");
            if (llm == IntPtr.Zero) throw new InvalidOperationException("LLM instance is not initialized");
            if (llamaLib.LLM_Status_Code() != 0)
            {
                string status_msg = Marshal.PtrToStringAnsi(llamaLib.LLM_Status_Message()) ?? string.Empty;
                throw new AccessViolationException(status_msg);
            }
        }

        public virtual void Dispose() { }

        ~LLM()
        {
            Dispose();
        }

        public string GetTemplate()
        {
            CheckLlamaLib();
            IntPtr result = llamaLib.LLM_Get_Template(llm);
            return Marshal.PtrToStringAnsi(result) ?? string.Empty;
        }

        public string ApplyTemplate(JArray messages = null)
        {
            if (messages == null)
                throw new ArgumentNullException(nameof(messages));
            CheckLlamaLib();
            IntPtr result = llamaLib.LLM_Apply_Template(llm, messages.ToString());
            return Marshal.PtrToStringAnsi(result) ?? string.Empty;
        }

        public List<int> Tokenize(string content)
        {
            if (string.IsNullOrEmpty(content))
                throw new ArgumentNullException(nameof(content));

            CheckLlamaLib();
            IntPtr result = llamaLib.LLM_Tokenize(llm, content);
            string resultStr = Marshal.PtrToStringAnsi(result) ?? string.Empty;
            List<int> ret = new List<int>();
            try
            {
                JArray json = JArray.Parse(resultStr);
                ret = json?.ToObject<List<int>>();
            }
            catch { }
            return ret;
        }

        public string Detokenize(List<int> tokens)
        {
            if (tokens == null)
                throw new ArgumentNullException(nameof(tokens));

            CheckLlamaLib();
            JArray tokensJSON = JArray.FromObject(tokens);
            IntPtr result = llamaLib.LLM_Detokenize(llm, tokensJSON.ToString());
            return Marshal.PtrToStringAnsi(result) ?? string.Empty;
        }

        public string Detokenize(int[] tokens)
        {
            if (tokens == null)
                throw new ArgumentNullException(nameof(tokens));
            return Detokenize(new List<int>(tokens));
        }

        public List<float> Embeddings(string content)
        {
            if (string.IsNullOrEmpty(content))
                throw new ArgumentNullException(nameof(content));

            CheckLlamaLib();

            IntPtr result = llamaLib.LLM_Embeddings(llm, content);
            string resultStr = Marshal.PtrToStringAnsi(result) ?? string.Empty;
            
            List<float> ret = new List<float>();
            try
            {
                JArray json = JArray.Parse(resultStr);
                ret = json?.ToObject<List<float>>();
            }
            catch { }
            return ret;
        }
     
        public string BuildParametersJSON(JObject parameters = null)
        {
            var json = new JObject
            {
                ["seed"] = seed,
                ["n_predict"] = numPredict,
                ["n_keep"] = numKeep,
                ["temperature"] = temperature,
            };
            if (jsonSchema != "") json["json_schema"] = jsonSchema;
            if (grammar != "") json["grammar"] = grammar;
            if (parameters != null)
            {
                foreach (var param in parameters)
                {
                    json[param.Key] = param.Value;
                }
            }
            return json.ToString();
        }

        public void CheckCompletionInternal(string prompt)
        {
            if (string.IsNullOrEmpty(prompt))
                throw new ArgumentNullException(nameof(prompt));
            CheckLlamaLib();
        }

        public string CompletionInternal(string prompt, LlamaLib.CharArrayCallback callback, int idSlot, JObject parameters, bool callbackWithJSON)
        {
            IntPtr result;
            if (callbackWithJSON) result = llamaLib.LLM_Completion_JSON(llm, prompt, callback, idSlot, BuildParametersJSON(parameters));
            else result = llamaLib.LLM_Completion(llm, prompt, callback, idSlot, BuildParametersJSON(parameters));
            return Marshal.PtrToStringAnsi(result) ?? string.Empty;
        }

        public string Completion(string prompt, LlamaLib.CharArrayCallback callback = null, int idSlot = -1, JObject parameters = null)
        {
            CheckCompletionInternal(prompt);
            return CompletionInternal(prompt, callback, idSlot, parameters, false);
        }

        public async Task<string> CompletionAsync(string prompt, LlamaLib.CharArrayCallback callback = null, int idSlot = -1, JObject parameters = null)
        {
            CheckCompletionInternal(prompt);
            return await Task.Run(() => CompletionInternal(prompt, callback, idSlot, parameters, false));
        }

        public JObject CompletionJSON(string prompt, LlamaLib.CharArrayCallback callback = null, int idSlot = -1, JObject parameters = null)
        {
            CheckCompletionInternal(prompt);
            return JObject.Parse(CompletionInternal(prompt, callback, idSlot, parameters, true));
        }

        public async Task<JObject> CompletionJSONAsync(string prompt, LlamaLib.CharArrayCallback callback = null, int idSlot = -1, JObject parameters = null)
        {
            CheckCompletionInternal(prompt);
            return await Task.Run(() => JObject.Parse(CompletionInternal(prompt, callback, idSlot, parameters, true)));
        }
    }

    // LLMLocal class
    public abstract class LLMLocal : LLM
    {
        protected LLMLocal() : base() { }

        protected LLMLocal(LlamaLib llamaLibInstance) : base(llamaLibInstance) { }

        public string Slot(int idSlot, string action, string filepath)
        {
            if (string.IsNullOrEmpty(action))
                throw new ArgumentNullException(nameof(action));
            if (string.IsNullOrEmpty(filepath))
                throw new ArgumentNullException(nameof(filepath));

            IntPtr result = llamaLib.LLM_Slot(llm, idSlot, action, filepath);
            return Marshal.PtrToStringAnsi(result) ?? string.Empty;
        }

        // Cancel methods
        public void Cancel(int idSlot)
        {
            CheckLlamaLib();
            llamaLib.LLM_Cancel(llm, idSlot);
        }
    }

    // LLMProvider class
    public abstract class LLMProvider : LLMLocal
    {
        protected LLMProvider() : base() { }

        protected LLMProvider(LlamaLib llamaLibInstance) : base(llamaLibInstance) { }


        public void SetTemplate(string template)
        {
            if (string.IsNullOrEmpty(template))
                throw new ArgumentNullException(nameof(template));
            CheckLlamaLib();
            llamaLib.LLM_Set_Template(llm, template);
        }

        // LoRA Weight methods
        public string BuildLoraWeightJSON(List<LoraIdScale> loras)
        {
            var jsonArray = new JArray();
            foreach (var lora in loras)
            {
                jsonArray.Add(new JObject { ["id"] = lora.Id, ["scale"] = lora.Scale });
            }
            return jsonArray.ToString();
        }

        public bool LoraWeight(List<LoraIdScale> loras)
        {
            if (loras == null)
                throw new ArgumentNullException(nameof(loras));
            
            var lorasJSON = BuildLoraWeightJSON(loras);
            return llamaLib.LLM_Lora_Weight(llm, lorasJSON);
        }

        public bool LoraWeight(params LoraIdScale[] loras)
        {
            return LoraWeight(new List<LoraIdScale>(loras));
        }

        // LoRA List methods
        public List<LoraIdScalePath> ParseLoraListJSON(string result)
        {
            var loras = new List<LoraIdScalePath>();
            try
            {
                var jsonArray = JArray.Parse(result);
                foreach (var item in jsonArray)
                {
                    int id = item["id"]?.ToObject<int>() ?? -1;
                    if (id < 0) continue;
                    loras.Add(new LoraIdScalePath(
                        id,
                        item["scale"]?.ToObject<float>() ?? 0.0f,
                        item["path"]?.ToString() ?? string.Empty
                    ));
                }
            }
            catch{ }
            return loras;
        }

        public string LoraListJSON()
        {
            CheckLlamaLib();
            var result = llamaLib.LLM_Lora_List(llm);
            return Marshal.PtrToStringAnsi(result) ?? string.Empty;
        }

        public List<LoraIdScalePath> LoraList()
        {
            var jsonResult = LoraListJSON();
            return ParseLoraListJSON(jsonResult);
        }

        // Server methods
        public bool Start()
        {
            CheckLlamaLib();
            llamaLib.LLM_Start(llm);
            return llamaLib.LLM_Started(llm);
        }

        public async Task<bool> StartAsync()
        {
            CheckLlamaLib();
            return await Task.Run(() => {
                llamaLib.LLM_Start(llm);
                return llamaLib.LLM_Started(llm);
            });
        }

        public bool Started()
        {
            CheckLlamaLib();
            return llamaLib.LLM_Started(llm);
        }

        public void Stop()
        {
            CheckLlamaLib();
            llamaLib.LLM_Stop(llm);
        }

        public void StartServer(string host = "0.0.0.0", int port = 0, string apiKey = "")
        {
            CheckLlamaLib();
            if (string.IsNullOrEmpty(host))
                host = "0.0.0.0";
            
            llamaLib.LLM_Start_Server(llm, host, port, apiKey ?? string.Empty);
        }

        public void StopServer()
        {
            CheckLlamaLib();
            llamaLib.LLM_Stop_Server(llm);
        }

        public void JoinService()
        {
            CheckLlamaLib();
            llamaLib.LLM_Join_Service(llm);
        }

        public void JoinServer()
        {
            CheckLlamaLib();
            llamaLib.LLM_Join_Server(llm);
        }

        public void SetSSL(string sslCert, string sslKey)
        {
            if (string.IsNullOrEmpty(sslCert))
                throw new ArgumentNullException(nameof(sslCert));
            if (string.IsNullOrEmpty(sslKey))
                throw new ArgumentNullException(nameof(sslKey));
            
            CheckLlamaLib();
            llamaLib.LLM_Set_SSL(llm, sslCert, sslKey);
        }

        public int EmbeddingSize()
        {
            CheckLlamaLib();
            return llamaLib.LLM_Embedding_Size(llm);
        }

        public override void Dispose()
        {
            lock (_disposeLock)
            {
                if (!disposed)
                {
                    if (llm != IntPtr.Zero && llamaLib != null)
                    {
                        try
                        {
                            llamaLib.LLM_Stop_Server(llm);
                            llamaLib.LLM_Stop(llm);
                            llamaLib.LLM_Delete(llm);
                        }
                        catch (Exception){ }
                    }
                    llamaLib?.Dispose();
                    llamaLib = null;
                    llm = IntPtr.Zero;
                }
                disposed = true;
            }
        }
    }

    // LLMService class
    public class LLMService : LLMProvider
    {
        public LLMService(string modelPath, int numThreads = -1, int numGpuLayers = 0, 
            int numParallel = 1, bool flashAttention = false, int contextSize = 4096, 
            int batchSize = 2048, bool embeddingOnly = false, string[] loraPaths = null)
        {
            if (string.IsNullOrEmpty(modelPath))
                throw new ArgumentNullException(nameof(modelPath));
            if (!File.Exists(modelPath))
                throw new FileNotFoundException($"Model file not found: {modelPath}");

            try
            {
                llamaLib = new LlamaLib(numGpuLayers > 0);
                llm = CreateLLM(modelPath, numThreads, numGpuLayers, numParallel, 
                    flashAttention, contextSize, batchSize, embeddingOnly, loraPaths);
            }
            catch
            {
                llamaLib?.Dispose();
                throw;
            }
        }

        public LLMService(LlamaLib llamaLibInstance, IntPtr llmInstance)
        {
            if (llamaLibInstance == null) throw new ArgumentNullException(nameof(llamaLibInstance));
            if (llmInstance == IntPtr.Zero) throw new ArgumentNullException(nameof(llmInstance));
            llamaLib = llamaLibInstance;
            llm = llmInstance;
        }

        public static LLMService FromCommand(string paramsString)
        {
            if (string.IsNullOrEmpty(paramsString))
                throw new ArgumentNullException(nameof(paramsString));

            LlamaLib llamaLibInstance = null;
            IntPtr llmInstance = IntPtr.Zero;
            try
            {
                llamaLibInstance = new LlamaLib(LlamaLib.Has_GPU_Layers(paramsString));
                llmInstance = llamaLibInstance.LLMService_From_Command(paramsString);
            }
            catch
            {
                llamaLibInstance?.Dispose();
                throw;
            }
            return new LLMService(llamaLibInstance, llmInstance);
        }

        private IntPtr CreateLLM(string modelPath, int numThreads, int numGpuLayers,
            int numParallel, bool flashAttention, int contextSize, int batchSize,
            bool embeddingOnly, string[] loraPaths)
        {
            IntPtr loraPathsPtr = IntPtr.Zero;
            int loraPathCount = 0;

            if (loraPaths != null && loraPaths.Length > 0)
            {
                loraPathCount = loraPaths.Length;
                // Allocate array of string pointers
                loraPathsPtr = Marshal.AllocHGlobal(IntPtr.Size * loraPathCount);
                
                try
                {
                    for (int i = 0; i < loraPathCount; i++)
                    {
                        if (string.IsNullOrEmpty(loraPaths[i]))
                            throw new ArgumentException($"Lora path at index {i} is null or empty");
                        
                        IntPtr stringPtr = Marshal.StringToHGlobalAnsi(loraPaths[i]);
                        Marshal.WriteIntPtr(loraPathsPtr, i * IntPtr.Size, stringPtr);
                    }
                }
                catch
                {
                    // Clean up if allocation failed
                    for (int i = 0; i < loraPathCount; i++)
                    {
                        IntPtr stringPtr = Marshal.ReadIntPtr(loraPathsPtr, i * IntPtr.Size);
                        if (stringPtr != IntPtr.Zero)
                            Marshal.FreeHGlobal(stringPtr);
                    }
                    Marshal.FreeHGlobal(loraPathsPtr);
                    throw;
                }
            }

            try
            {
                var llm = llamaLib.LLMService_Construct(
                    modelPath, numThreads, numGpuLayers, numParallel,
                    flashAttention, contextSize, batchSize, embeddingOnly,
                    loraPathCount, loraPathsPtr);

                if (llm == IntPtr.Zero)
                    throw new InvalidOperationException("Failed to create LLMService");

                return llm;
            }
            finally
            {
                // Clean up allocated strings
                if (loraPathsPtr != IntPtr.Zero)
                {
                    for (int i = 0; i < loraPathCount; i++)
                    {
                        IntPtr stringPtr = Marshal.ReadIntPtr(loraPathsPtr, i * IntPtr.Size);
                        if (stringPtr != IntPtr.Zero)
                            Marshal.FreeHGlobal(stringPtr);
                    }
                    Marshal.FreeHGlobal(loraPathsPtr);
                }
            }
        }
    }

    // LLMClient class
    public class LLMClient : LLMLocal
    {
        public LLMClient(LLMProvider provider)
        {
            if (provider.disposed)
                throw new ObjectDisposedException(nameof(provider));

            llamaLib = provider.llamaLib;
            llm = CreateClient(provider);
        }

        public LLMClient(string url, int port)
        {
            if (string.IsNullOrEmpty(url))
                throw new ArgumentNullException(nameof(url));

            try
            {
                llamaLib = new LlamaLib(false);
                llm = CreateRemoteClient(url, port);
            }
            catch
            {
                llamaLib?.Dispose();
                throw;
            }
        }

        private IntPtr CreateClient(LLMProvider provider)
        {
            var llm = llamaLib.LLMClient_Construct(provider.llm);
            if (llm == IntPtr.Zero)
                throw new InvalidOperationException("Failed to create LLMClient");
            return llm;
        }

        private IntPtr CreateRemoteClient(string url, int port)
        {
            var llm = llamaLib.LLMClient_Construct_Remote(url, port);
            if (llm == IntPtr.Zero)
                throw new InvalidOperationException($"Failed to create remote LLMClient for {url}:{port}");
            return llm;
        }
    }
}